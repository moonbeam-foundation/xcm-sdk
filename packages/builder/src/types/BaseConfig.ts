export interface BaseConfigConstructorParams {
  module: string;
  func: string;
}

export class BaseConfig {
  readonly module: string;

  readonly func: string;

  constructor({ module, func }: BaseConfigConstructorParams) {
    this.module = module;
    this.func = func;
  }
}
