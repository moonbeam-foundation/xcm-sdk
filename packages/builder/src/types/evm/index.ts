export * from './ContractConfig';
export * from './EvmQueryConfig';
