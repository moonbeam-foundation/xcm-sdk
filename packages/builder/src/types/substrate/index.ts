export * from './ExtrinsicConfig';
export * from './SubstrateCallConfig';
export * from './SubstrateQueryConfig';
