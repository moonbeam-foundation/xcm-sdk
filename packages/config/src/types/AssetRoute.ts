import type {
  AssetMinConfigBuilder,
  BalanceConfigBuilder,
  ContractConfigBuilder,
  ExtrinsicConfigBuilder,
  FeeConfigBuilder,
} from '@moonbeam-network/xcm-builder';
import type {
  AnyChain,
  Asset,
  ChainAsset,
  SetOptional,
} from '@moonbeam-network/xcm-types';

export interface AssetRouteConstructorParams {
  source: SourceConfig;
  destination: DestinationConfig;
  contract?: ContractConfigBuilder;
  extrinsic?: ExtrinsicConfigBuilder;
}

export interface SourceConfig {
  asset: Asset;
  chain: AnyChain;
  balance: BalanceConfigBuilder;
  fee: FeeConfig;
  destinationFee?: {
    asset?: Asset;
    balance: BalanceConfigBuilder;
  };
  min?: AssetMinConfigBuilder;
}

export interface DestinationConfig extends Omit<SourceConfig, 'fee'> {
  fee: DestinationFeeConfig;
}

export interface FeeConfig {
  asset: Asset;
  balance: BalanceConfigBuilder;
  // NOTE: Sometimes we need to add some extra amount ("XCM Delivery Fee") to a fee
  // that is returned by "paymentInfo" for extrinsic to not fail.
  extra?: number;
}

export interface DestinationFeeConfig
  extends SetOptional<FeeConfig, 'balance'> {
  amount: number | FeeConfigBuilder;
}

export class AssetRoute {
  readonly source: SourceConfig;

  readonly destination: DestinationConfig;

  readonly contract?: ContractConfigBuilder;

  readonly extrinsic?: ExtrinsicConfigBuilder;

  constructor({
    source,
    destination,
    contract,
    extrinsic,
  }: AssetRouteConstructorParams) {
    this.source = source;
    this.destination = destination;
    this.contract = contract;
    this.extrinsic = extrinsic;
  }

  getDestinationFeeAssetOnSource(): ChainAsset {
    return this.source.chain.getChainAsset(
      this.source.destinationFee?.asset || this.destination.fee.asset,
    );
  }
}
