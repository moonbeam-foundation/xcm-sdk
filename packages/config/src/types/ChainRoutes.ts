import type { AnyAsset, AnyChain, Asset } from '@moonbeam-network/xcm-types';
import { getKey } from '../config.utils';
import {
  AssetRoute,
  type AssetRouteConstructorParams,
  type SourceConfig,
} from './AssetRoute';

export interface ChainRoutesConstructorParams {
  chain: AnyChain;
  routes: RoutesParam[];
}

interface RoutesParam extends Omit<AssetRouteConstructorParams, 'source'> {
  source: Omit<SourceConfig, 'chain'>;
}

export class ChainRoutes {
  readonly chain: AnyChain;

  protected routes: Map<string, AssetRoute>;

  constructor({ chain, routes }: ChainRoutesConstructorParams) {
    this.chain = chain;
    this.routes = new Map(
      routes.map(({ source, destination, contract, extrinsic, monitoring }) => [
        `${source.asset.key}-${destination.chain.key}`,
        new AssetRoute({
          source: { ...source, chain },
          destination,
          contract,
          extrinsic,
          monitoring,
        }),
      ]),
    );
  }

  getRoutes(): AssetRoute[] {
    return Array.from(this.routes.values());
  }

  getAssetRoutes(keyOrAsset: string | AnyAsset): AssetRoute[] {
    const key = getKey(keyOrAsset);

    return this.getRoutes().filter((route) => route.source.asset.key === key);
  }

  getAssetDestinations(keyOrAsset: string | AnyAsset): AnyChain[] {
    return this.getAssetRoutes(keyOrAsset).map(
      (assetConfig) => assetConfig.destination.chain,
    );
  }

  getDestinationAssets(keyOrChain: string | AnyChain): Asset[] {
    const key = getKey(keyOrChain);

    return this.getRoutes()
      .filter((route) => route.destination.chain.key === key)
      .map((route) => route.source.asset);
  }

  getAssetRoute(
    asset: string | AnyAsset,
    destination: string | AnyChain,
  ): AssetRoute {
    const assetKey = getKey(asset);
    const destKey = getKey(destination);
    const route = this.routes.get(`${assetKey}-${destKey}`);

    if (!route) {
      throw new Error(
        `AssetRoute for asset ${assetKey} and destination ${destKey} not found`,
      );
    }

    return route;
  }
}
