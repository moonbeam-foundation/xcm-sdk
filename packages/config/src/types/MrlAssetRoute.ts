import type { BalanceConfigBuilder } from '@moonbeam-network/xcm-builder';
import type { Asset } from '@moonbeam-network/xcm-types';
import {
  AssetRoute,
  type AssetRouteConstructorParams,
  type MrlConfig,
  type SourceConfig,
} from './AssetRoute';

export interface MrlAssetRouteConstructorParams
  extends AssetRouteConstructorParams {
  source: MrlSourceConfig;
  mrl: MrlConfig;
}

export interface MrlSourceConfig extends SourceConfig {
  moonChainFee?: {
    asset: Asset;
    balance: BalanceConfigBuilder;
  };
}

export class MrlAssetRoute extends AssetRoute {
  readonly mrl: MrlConfig;
  readonly source: MrlSourceConfig;

  constructor({
    source,
    destination,
    contract,
    extrinsic,
    mrl,
  }: MrlAssetRouteConstructorParams & { source: MrlSourceConfig }) {
    super({ source, destination, contract, extrinsic });
    this.mrl = mrl;
    this.source = source;
  }
}
