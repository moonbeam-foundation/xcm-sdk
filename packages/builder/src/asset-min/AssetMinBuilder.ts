import type { Option } from '@polkadot/types';
import type { PalletAssetsAssetDetails } from '@polkadot/types/lookup';
import { getExtrinsicAccount } from '../extrinsic/ExtrinsicBuilder.utils';
import { SubstrateQueryConfig } from '../types/substrate/SubstrateQueryConfig';
import type { AssetMinConfigBuilder } from './AssetMinBuilder.interfaces';

export function AssetMinBuilder() {
  return {
    assetRegistry,
    assets,
    foreignAssets,
  };
}

function assetRegistry() {
  const pallet = 'assetRegistry';
  return {
    assetMetadatas: (): AssetMinConfigBuilder => ({
      build: ({ asset }) =>
        new SubstrateQueryConfig({
          module: pallet,
          func: 'assetMetadatas',
          args: [asset],
          // biome-ignore lint/suspicious/noExplicitAny: not sure how to fix this
          transform: async (response: Option<any>): Promise<bigint> =>
            response.unwrapOrDefault().minimalBalance.toBigInt(),
        }),
    }),
    currencyMetadatas: (): AssetMinConfigBuilder => ({
      build: ({ asset }) =>
        new SubstrateQueryConfig({
          module: pallet,
          func: 'currencyMetadatas',
          args: [asset],
          // biome-ignore lint/suspicious/noExplicitAny: not sure how to fix this
          transform: async (response: Option<any>): Promise<bigint> =>
            response.unwrapOrDefault().minimalBalance.toBigInt(),
        }),
    }),
  };
}

function assets() {
  return {
    asset: (): AssetMinConfigBuilder => ({
      build: ({ asset }) =>
        new SubstrateQueryConfig({
          module: 'assets',
          func: 'asset',
          args: [asset],
          transform: async (
            response: Option<PalletAssetsAssetDetails>,
          ): Promise<bigint> =>
            response.unwrapOrDefault().minBalance.toBigInt(),
        }),
    }),
  };
}

function foreignAssets() {
  return {
    asset: (): AssetMinConfigBuilder => ({
      build: ({ address }) => {
        if (!address) {
          throw new Error(
            'Asset address is missing for foreignAssets.asset min calculation',
          );
        }
        const multilocation = {
          parents: 2,
          interior: {
            X2: [
              {
                globalconsensus: {
                  ethereum: {
                    chainId: 1,
                  },
                },
              },
              getExtrinsicAccount(address),
            ],
          },
        };
        return new SubstrateQueryConfig({
          module: 'foreignAssets',
          func: 'asset',
          args: [multilocation],
          transform: async (
            response: Option<PalletAssetsAssetDetails>,
          ): Promise<bigint> =>
            response.unwrapOrDefault().minBalance.toBigInt(),
        });
      },
    }),
  };
}
