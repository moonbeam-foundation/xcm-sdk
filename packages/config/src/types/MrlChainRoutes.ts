import type { AnyAsset, AnyChain } from '@moonbeam-network/xcm-types';
import { getKey } from '../config.utils';
import { ChainRoutes, type ChainRoutesConstructorParams } from './ChainRoutes';
import {
  MrlAssetRoute,
  type MrlAssetRouteConstructorParams,
  type MrlSourceConfig,
} from './MrlAssetRoute';

export interface MrlChainRoutesConstructorParams
  extends ChainRoutesConstructorParams {
  routes: MrlRoutesParam[];
}

interface MrlRoutesParam
  extends Omit<MrlAssetRouteConstructorParams, 'source'> {
  source: Omit<MrlSourceConfig, 'chain'>;
}

export class MrlChainRoutes extends ChainRoutes {
  readonly #routes: Map<string, MrlAssetRoute>;

  constructor({ chain, routes }: MrlChainRoutesConstructorParams) {
    super({ chain, routes });
    this.#routes = new Map(
      routes.map(({ source, destination, contract, extrinsic, mrl }) => [
        `${source.asset.key}-${destination.chain.key}`,
        new MrlAssetRoute({
          source: { ...source, chain },
          destination,
          contract,
          extrinsic,
          mrl,
        }),
      ]),
    );
  }

  getRoutes(): MrlAssetRoute[] {
    return Array.from(this.#routes.values());
  }

  getAssetRoute(
    asset: string | AnyAsset,
    destination: string | AnyChain,
  ): MrlAssetRoute {
    const assetKey = getKey(asset);
    const destKey = getKey(destination);
    const route = this.#routes.get(`${assetKey}-${destKey}`);

    if (!route) {
      throw new Error(
        `AssetRoute for asset ${assetKey} and destination ${destKey} not found`,
      );
    }

    return route;
  }
}
