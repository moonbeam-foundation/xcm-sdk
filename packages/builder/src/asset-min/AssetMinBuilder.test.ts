import { describe, expect, it } from 'vitest';

import { TypeRegistry, U128 } from '@polkadot/types';
import { AssetMinBuilder } from './AssetMinBuilder';

function balanceOf(number: number | string): U128 {
  return new U128(new TypeRegistry(), number);
}

/**
 * Using snapshot to test bigint values because jest does not support bigint
 */

describe('assetMinBuilder', () => {
  const asset = '<ASSET>';

  describe('assetRegistry', () => {
    describe('assetMetadatas', () => {
      const config = AssetMinBuilder()
        .assetRegistry()
        .assetMetadatas()
        .build({ asset });

      it('should be correct config', () => {
        expect(config).toMatchSnapshot();
      });

      it('should transform correctly', async () => {
        await expect(
          config.transform({
            unwrapOrDefault: () => ({
              minimalBalance: balanceOf(999),
            }),
          }),
        ).resolves.toMatchSnapshot();
      });
    });

    describe('currencyMetadatas', () => {
      const config = AssetMinBuilder()
        .assetRegistry()
        .currencyMetadatas()
        .build({ asset });

      it('should be correct config', () => {
        expect(config).toMatchSnapshot();
      });

      it('should transform correctly', async () => {
        await expect(
          config.transform({
            unwrapOrDefault: () => ({
              minimalBalance: balanceOf(999),
            }),
          }),
        ).resolves.toMatchSnapshot();
      });
    });
  });

  describe('assets', () => {
    describe('asset', () => {
      const config = AssetMinBuilder().assets().asset().build({ asset });

      it('should be correct config', () => {
        expect(config).toMatchSnapshot();
      });

      it('should transform correctly', async () => {
        await expect(
          config.transform({
            unwrapOrDefault: () => ({
              minBalance: balanceOf(999),
            }),
          }),
        ).resolves.toMatchSnapshot();
      });
    });
  });

  describe('foreignAssets', () => {
    describe('asset', () => {
      const config = AssetMinBuilder().foreignAssets().asset().build({
        asset,
        address: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
      });

      it('should be correct config', () => {
        expect(config).toMatchSnapshot();
      });

      it('should transform correctly', async () => {
        await expect(
          config.transform({
            unwrapOrDefault: () => ({
              minBalance: balanceOf(999),
            }),
          }),
        ).resolves.toMatchSnapshot();
      });
    });
  });
});
