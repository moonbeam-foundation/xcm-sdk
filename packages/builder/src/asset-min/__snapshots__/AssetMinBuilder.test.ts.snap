// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`assetMinBuilder > assetRegistry > assetMetadatas > should be correct config 1`] = `
SubstrateQueryConfig {
  "args": [
    "<ASSET>",
  ],
  "func": "assetMetadatas",
  "module": "assetRegistry",
  "transform": [Function],
}
`;

exports[`assetMinBuilder > assetRegistry > assetMetadatas > should transform correctly 1`] = `999n`;

exports[`assetMinBuilder > assetRegistry > currencyMetadatas > should be correct config 1`] = `
SubstrateQueryConfig {
  "args": [
    "<ASSET>",
  ],
  "func": "currencyMetadatas",
  "module": "assetRegistry",
  "transform": [Function],
}
`;

exports[`assetMinBuilder > assetRegistry > currencyMetadatas > should transform correctly 1`] = `999n`;

exports[`assetMinBuilder > assets > asset > should be correct config 1`] = `
SubstrateQueryConfig {
  "args": [
    "<ASSET>",
  ],
  "func": "asset",
  "module": "assets",
  "transform": [Function],
}
`;

exports[`assetMinBuilder > assets > asset > should transform correctly 1`] = `999n`;

exports[`assetMinBuilder > foreignAssets > asset > should be correct config 1`] = `
SubstrateQueryConfig {
  "args": [
    {
      "interior": {
        "X2": [
          {
            "globalconsensus": {
              "ethereum": {
                "chainId": 1,
              },
            },
          },
          {
            "AccountKey20": {
              "key": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
            },
          },
        ],
      },
      "parents": 2,
    },
  ],
  "func": "asset",
  "module": "foreignAssets",
  "transform": [Function],
}
`;

exports[`assetMinBuilder > foreignAssets > asset > should transform correctly 1`] = `999n`;
